corehq/apps/registration/tests/test_utils.py

```diff
@@ -32,98 +32,91 @@ def _noop(*args, **kwargs):
     pass
 
 
+@mock.patch('corehq.apps.registration.utils.notify_exception', _noop)
 class TestRequestNewDomain(TestCase):
 
     @classmethod
     def setUpClass(cls):
         super().setUpClass()
-        cls.new_user = WebUser.create(
-            None, 'new@dimagi.org', 'testpwd', None, None
-        )
+        cls.username = 'new@dimagi.org'
         cls.request = RequestFactory().get('/registration')
-        django_user = User.objects.get(username=cls.new_user.username)
-        cls.request.user = django_user
         cls.domain_sso_test = 'test-sso-1'
         cls.domain_test = 'test-1'
 
+    def setUp(self):
+        def delete_user_if_exists():
+            user = WebUser.get_by_username(self.username)
+            if user is not None:
+                # normally it will be deleted when the domain is deleted
+                user.delete(None, None)
+        self.new_user = WebUser.create(None, self.username, 'testpwd', None, None)
+        self.request.user = User.objects.get(username=self.username)
+        self.addClassCleanup(delete_user_if_exists)
+
     def tearDown(self):
         subscribed_domains = [self.domain_sso_test, self.domain_test]
         for domain in subscribed_domains:
-            Subscription._get_active_subscription_by_domain.clear(
-                Subscription,
-                domain
-            )
-        SelfSignupWorkflow.objects.filter(domain__in=subscribed_domains).delete()
-
-        failed_subscribed_domains = ['subscription-failed', 'init-default-roles-failed']
-        for test_domain in failed_subscribed_domains:
-            domain = Domain.get_by_name(test_domain)
-            if domain is not None:
-                domain.delete()
-
+            Subscription._get_active_subscription_by_domain.clear(Subscription, domain)
+            SelfSignupWorkflow.get_in_progress_for_domain.clear(SelfSignupWorkflow, domain)
         super().tearDown()
 
-    @classmethod
-    def tearDownClass(cls):
-        cls.new_user.delete(cls.domain_test, deleted_by=None)
-        super().tearDownClass()
-
     # if we don't patch the following, NoBrokersAvailable is thrown due to publish_domain_saved
     @mock.patch('corehq.apps.registration.utils._setup_subscription', _noop)
-    @mock.patch('corehq.apps.registration.utils.notify_exception', _noop)
     def test_domain_is_active_for_new_sso_user(self):
         """
         Ensure that the first domain created by a new SSO user is active.
         """
+        self.addCleanup(_cleanup_domain, 'test-sso-1')
```

Comment:
Why would we need this line? It seems like the only cleanup necessary would be for domain name returned from "request_new_domain", which would cover this. Same goes for the other test.

Reply:
That's true, although if a domain named "test-sso-1" did exist prior to this test being run, then the name returned by `request_new_domain()` would be different, and that would imply that some other test did not clean up properly.

I think this doesn't hurt anything, although could see an argument for asserting that `domain_name == "test-sso-1"` instead to fail immediately when it is discovered that some other test did not clean up properly. I don't love the idea of silently ignoring unexpected database state.

Reply:
Yeah I like an assertion better too, but am fine either way.